### PR TITLE
mvapich2: added version 2.3rc1 + variant to use alloca if available

### PR DESCRIPTION
**Pull request by alalazo**
Besides this small update, I have a question. `mvapich2@1.9` (the oldest version that is checksummed) is not downloadable anymore. I was wondering if we should start having deprecation policies for software or not. I would be for removing the version and clean the recipe from everything that was `@1.9` specific, but would like to hear before what people think about it.

### CODE REVIEW

**Comment by davydden** [comment]
> I would be for removing the version and clean the recipe from everything that was @1.9 specific

👍

File: var/spack/repos/builtin/packages/mvapich2/package.py [hunk 1]
```diff
@@ -113,10 +115,7 @@ class Mvapich2(AutotoolsPackage):
 
     def url_for_version(self, version):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mvapich2/package.py]: This function can be removed altogether if you're getting rid of 1.9.

**Comment by healther** [comment]
Question: Should this be general policy? I.e. we remove versions from `package.py` once they are no longer available online? What about spack-maintained mirrors?

**Comment by alalazo** [comment]
> Should this be general policy? I.e. we remove versions from package.py once they are no longer available online? What about spack-maintained mirrors?

I'll leave the PR open for a few more days, so that people could comment on this. 

I'd be for removing old versions of packages, in particular considering that lately we started releasing  new versions of Spack more frequently. I think it's quite open to discussion whether to remove them based on their availability on the original site or on some other criteria,  and probably should vary among packages and be guided by common sense, rather than strict rules.

In any case, if people decide to never delete a package version (is this sustainable in the long run?) or if it seems too early to delete `mvapich2@1.9` I can just rebase this PR and skip the last commit.

**Comment by 0xaf1f** [comment]
I find it quite strange that one file is used to define builds of every version of a package. Is it out of the question to use one file per package-version?

**Comment by adamjstewart** [comment]
@0xaf1f That's how package managers like Homebrew and Easybuild work, but I've never seen the attraction. For 99% of packages, the installation instructions are identical for every version. For the other 1%, it's usually a 1 or 2 line if-statement. I don't see why anyone would want to maintain and update 10 different versions of the same `package.py`.

**Comment by 0xaf1f** [comment]
So then maybe I misunderstood the problem you're talking about. From what I understood on the slack replies following the link to this thread, there was mention of complexity increases in `package.py` due to having to support old versions in the same script. Are you here talking about `package.py` getting too long because of a long list of checksums?

**Comment by adamjstewart** [comment]
There is some increasing complexity, but it's manageable. Certainly fewer lines of code than a `package.py` for every version. The only time the complexity blows up is when a package changes from one build system to another. We currently don't have a great way to handle that. For example, `metis` switched from Make to CMake while `arpack-ng` switched from Autotools to CMake.

The question here is whether the increased complexity (a few lines of code) is worth keeping around an ancient version that can't even be downloaded anymore.

**Comment by 0xaf1f** [comment]
Even if the version is not ancient, the packaging can change a few times between modern versions and still present a similar situation. Even if the build system doesn't change, maybe they take a new configure option that they didn't take before, etc.

You'd have to stop supporting old versions at some point, but do you want the decision to be influenced by a situation where developers have been tweaking their build system too many times in the past year?

Anyway, if it's just a few lines of code as you said, do you really consider all this a problem? I'm seeing mixed messages about this here.

And I wanted to comment on this point:
> I don't see why anyone would want to maintain and update 10 different versions of the same package.py.

For the most part, old versions generally need little or no maintenance since they're not changing anymore. Maybe this isn't true for Spack if the package format hasn't stabilized, so I see your point if the old versions do in fact need maintenance.

Right now, if you decide to get rid of packaging support for a version and someone later wants to dig it up, they'd have to go through the git history of the `package.py` file to see when the support for that version was dropped. If they were separate files, it would be trivial to find it. Debian actually has a dedicated repository per package and the repositories would be tagged and versioned.

With one file to package all versions, there's also an interesting class of bugs I haven't seen before like https://github.com/spack/spack/pull/6463 (which I ran into when using the latest release) where adding support for the new package broke the old one, when the old one would have probably been just fine for the versions it was written for if it was left untouched.

I'm not continuing to advocate for one file per package-version, I just wanted to explain how it does have upsides.

**Comment by healther** [comment]
The one feature that is currently missing is a way to reflect build system changes, without having to reimplement the different install phases with manual if-checks for the versions. I.e. a way to say: 
```
version('x.y', build_system='Makefile')
version('x.z', build_system='CMake')
```
or
```
build_system('Makefile', when='@x.y:x.z')
build_system('CMake', when='@x.z:')
```
But that's really more a case of not having a significant need for it. 
In case of really overburding logic in a `package.py` (and the need to delete older versions) we might have to think about making legacy versions more easily available. The point is a bit academic as long as there is no public source mirror associated with spack, because right now it doesn't matter if we delete `1.9` or not as `spack` can't get at the sources anyways.

**Comment by alalazo** [comment]
This one has been pending for a while. Anybody that doesn't want the last commit merged?

**Pull request merged**